astropy/units/core.py

```diff
@@ -669,12 +670,14 @@ def _get_physical_type_id(self):
         this unit, without the scale.  Since it is hashable, it is
         useful as a dictionary key.
         """
-        unit = self.decompose()
-        r = zip([x.name for x in unit.bases], unit.powers)
-        # bases and powers are already sorted in a unique way
-        # r.sort()
-        r = tuple(r)
-        return r
+        if self._type_id is None:
+            unit = self.decompose()
+            r = zip([x.name for x in unit.bases], unit.powers)
+            # bases and powers are already sorted in a unique way
+            # r.sort()
+            self._type_id = tuple(r)
```

Comment:
This code will only be executed if `_type_id` isn't computed already, but replacing the list comprehension with a generator expression should be faster.
```suggestion
            self._type_id = tuple(zip((base.name for base in unit.bases), unit.powers))
```